driver/compile.go

```diff
@@ -27,7 +31,80 @@ type Config struct {
 	Warnings          chan string
 }
 
-func compile(ctx context.Context, program ast.Proc, zctx *resolver.Context, msrc MultiSource, mcfg MultiConfig) (*muxOutput, error) {
+func programPrep(program ast.Proc, sortKey string, sortReversed bool) (ast.Proc, ast.BooleanExpr, filter.Filter, error) {
+	ReplaceGroupByProcDurationWithKey(program)
+	if sortKey != "" {
+		setGroupByProcInputSortDir(program, sortKey, zbufDirInt(sortReversed))
+	}
+	var filterExpr ast.BooleanExpr
+	var filt filter.Filter
+	filterExpr, program = liftFilter(program)
+	if filterExpr != nil {
+		var err error
+		if filt, err = filter.Compile(filterExpr); err != nil {
+			return nil, nil, nil, err
+		}
+	}
+	return program, filterExpr, filt, nil
+}
+
+type scannerProc struct {
+	scanner.Scanner
+}
+
+func (s *scannerProc) Done() {}
+
+type namedScanner struct {
+	scanner.Scanner
+	name string
+}
+
+func (n *namedScanner) Pull() (zbuf.Batch, error) {
+	b, err := n.Scanner.Pull()
+	if err != nil {
+		err = fmt.Errorf("%s: %w", n.name, err)
+	}
+	return b, err
+}
+
+func compileSingle(ctx context.Context, program ast.Proc, zctx *resolver.Context, r zbuf.Reader, cfg Config) (*muxOutput, error) {
+	if cfg.Logger == nil {
+		cfg.Logger = zap.NewNop()
+	}
+	if cfg.Span.Dur == 0 {
+		cfg.Span = nano.MaxSpan
+	}
+	if cfg.Warnings == nil {
+		cfg.Warnings = make(chan string, 5)
+	}
+
+	program, filterExpr, filt, err := programPrep(program, cfg.ReaderSortKey, cfg.ReaderSortReverse)
+	if err != nil {
+		return nil, err
+	}
+
+	sn, err := scanner.NewScanner(ctx, r, filt, filterExpr, cfg.Span)
```

Comment:
```suggestion
	program, fil, filterExpr, err := programPrep(program, cfg.ReaderSortKey, cfg.ReaderSortReverse)
	if err != nil {
		return nil, err
	}

	sn, err := scanner.NewScanner(ctx, r, filt, filterExpr, cfg.Span)
```
Maybe return values in the same order they'll be passed to scanner.NewScanner?